### PR TITLE
Added support for MH mailboxes and several other features/improvements/fixes

### PR DESCRIPTION
**Pull request by ad-ast**
Hello,

this PR adds support for MH mailbox directories. While implementing that feature several other changes suddenly happened:

* added recursive multipart parsing
* more lenient with JSON parser exceptions (skip unparsable message)
* added option to filter for text content types only
* added option to skip x-* header indexing
* fixed delete index HTTP request (spurious refresh parameter)
* optimized message skipping (skip on keys to avoid unnecessary slow read operations for message content; this is useful for MH mailboxes)
* added option for dry running, i.e. not uploading to ES

All changes are backwards compatible. Existing use cases should work like before, no need to change parameters etc.

I made JSON parsing more lenient because I encountered several unparsable messages in my personal mailbox. I did not look into the cause very deeply, but on a first glance these were encoding errors made by mail clients. They happened mostly on older messages.

### CODE REVIEW

**Comment by oliver006** [comment]
Thanks for the PR!
I'll get to reviewing it in the next few days.

File: src/index_emails.py [hunk 1]
```diff
@@ -166,22 +180,24 @@ def load_from_file():
...
 
-    emailParser = DelegatingEmailParser([AmazonEmailParser(), SteamEmailParser()])
+    #emailParser = DelegatingEmailParser([AmazonEmailParser(), SteamEmailParser()])
 
-    for msg in mbox:
-        count += 1
-        if count < tornado.options.options.skip:
-            continue
+    #Skipping on keys to avoid expensive read operations on skipped messages
+    msgkeys = mbox.keys()[tornado.options.options.skip:]
```
Comment (oliver006) [inline src/index_emails.py:192]: This is neat.

File: src/index_emails.py [hunk 2]
```diff
@@ -166,22 +180,24 @@ def load_from_file():
...
-    mbox = mailbox.mbox(tornado.options.options.infile)
+    if tornado.options.options.infile:
+        logging.info("Starting import from mbox file %s" % tornado.options.options.infile)
+        mbox = mailbox.mbox(tornado.options.options.infile)
+    else:
+        logging.info("Starting import from MH directory %s" % tornado.options.options.indir)
+        mbox = mailbox.MH(tornado.options.options.indir, factory=None, create=False)
 
-    emailParser = DelegatingEmailParser([AmazonEmailParser(), SteamEmailParser()])
+    #emailParser = DelegatingEmailParser([AmazonEmailParser(), SteamEmailParser()])
```
Comment (oliver006) [inline src/index_emails.py]: Can you bring this back plz?

**Review by oliver006 — COMMENTED** [review]
Looks great, thanks for the PR!

Just a small comment to bring back what you commented out.

File: src/index_emails.py (hunk 2 above)
Comment (ad-ast) [reply src/index_emails.py]: Done.

Pylint complained about the unused variable, so I cleaned it up while in the vicinity.

Comment (oliver006) [reply src/index_emails.py]: Good point.
You know, let's just remove all the email parser stuff, don't think that's really relevant any longer. Do you mind removing the three files and few lines of code from this file ?

Comment (ad-ast) [reply src/index_emails.py]: Sure. Files and their usages are removed now.

File: src/index_emails.py [hunk 3]
```diff
@@ -39,9 +36,8 @@ def strip_html_css_js(msg):
 
 def delete_index():
     try:
-        url = "%s/%s?refresh=true" % (tornado.options.options.es_url, tornado.options.options.index_name)
+        url = "%s/%s" % (tornado.options.options.es_url, tornado.options.options.index_name)
```
Comment (oliver006) [inline src/index_emails.py:39]: Question: why did you remove the refresh=true here and in body below? (it's been a while since i've worked with ES and I've not kept up with newer versions)

Comment (ad-ast) [reply src/index_emails.py:39]: I'm using ES 7.7. When setting that parameter I'm getting a HTTP status code 400 with a root cause 'illegal_argument_exception' -> 'request [...] contains unrecognized parameter'.
The index API docs at https://www.elastic.co/guide/en/elasticsearch/reference/7.7/indices-delete-index.html also don't mention any refresh parameter.

On the other hand, the documents API seems to support such an parameter: https://www.elastic.co/guide/en/elasticsearch/reference/7.7/docs-refresh.html

Maybe this was changed in later ES versions?

Comment (oliver006) [reply src/index_emails.py:39]: Very possible. As I said, not really up to date with my ES knowledge.
Thanks for linking the docs, this should suffice for now as documentation.

**Comment by ad-ast** [comment]
Thanks for merging!
